aten/src/ATen/CPUApplyUtils.h

```diff
@@ -22,289 +24,224 @@ namespace at {
+  // An empty tensor has no elements
+  for (auto& t : tensors)
+    if (t.sizes().equals({0}))
+      return false;
+  return true;
 }
 
-/*
-  Apply a pointwise operator to three tensors.
+template <std::size_t I = 0, typename... Tp>
+inline typename std::enable_if<I == sizeof...(Tp), void>::type iterate_all(
+    std::tuple<Tp...>& t) {}
 
-  The calling convention for op is a function/functor that takes takes three references to
-  type scalar; at least one of these references should be non-const in order to write the output.
-  For example, to compute a = b + c, op would be of the form:
-  [](scalar &a_val, const scalar &b_val, const scalar &c_val) { a_val = b_val + c_val; };
-*/
-template<typename scalar1, typename scalar2, typename scalar3, typename Op>
-void CPU_tensor_apply3(Tensor tensor1, Tensor tensor2, Tensor tensor3, Op op) {
-  CPU_tensor_apply3_dim<scalar1, scalar2, scalar3, Op>(tensor1, tensor2, tensor3, -1, op);
+template <std::size_t I = 0, typename... Tp>
+    inline typename std::enable_if <
+    I<sizeof...(Tp), void>::type iterate_all(std::tuple<Tp...>& t) {
+  std::get<I>(t).iterate();
+  iterate_all<I + 1, Tp...>(t);
 }
 
-template <typename scalar1, typename scalar2, typename scalar3, typename scalar4, typename Op>
-void CPU_tensor_apply4_dim(Tensor &tensor1, Tensor& tensor2, Tensor& tensor3, Tensor& tensor4, int64_t dim, Op op) {
-  checkBackend("CPU_tensor_apply4", {tensor1, tensor2, tensor3, tensor4}, Backend::CPU);
-  bool TH_TENSOR_APPLY_hasFinished = false;
-  int64_t TH_TENSOR_dim_index = 0;
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar1, tensor1, dim, 1)
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar2, tensor2, dim, 1)
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar3, tensor3, dim, 1)
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar4, tensor4, dim, 1)
+template <std::size_t I = 0, typename... Tp>
+inline typename std::enable_if<I == sizeof...(Tp), bool>::type is_finished_all(
+    std::tuple<Tp...>& t) {
+  return true;
+}
 
-  int elements_equal = 1;
-  auto t1_numel = tensor1.numel();
-  auto t2_numel = tensor2.numel();
-  auto t3_numel = tensor3.numel();
-  auto t4_numel = tensor4.numel();
-  if(t1_numel!= t2_numel) {
-    elements_equal = 0;
-  } else if(t1_numel != t3_numel) {
-    elements_equal = 0;
-  } else if(t1_numel != t4_numel) {
-      elements_equal = 0;
-  }
-  if (elements_equal == 0) {
-    std::ostringstream oss;
-    oss << "inconsistent tensor size, expected " << tensor1.sizes() << ", " << tensor2.sizes() << ", "
-        << tensor3.sizes() << ", and " << tensor4.sizes() << " to have the same number of elements, but got "
-        << t1_numel << ", " << t2_numel << ", " << t3_numel << ", and " << t4_numel << " elements respectively";
-    throw std::runtime_error(oss.str());
-  }
+template <std::size_t I = 0, typename... Tp>
+    inline typename std::enable_if <
+    I<sizeof...(Tp), bool>::type is_finished_all(std::tuple<Tp...>& t) {
+  return std::get<I>(t).is_finished() && is_finished_all<I + 1, Tp...>(t);
+}
```

Comment:
You can implement both of those functions without iterating over tuples. It's much simpler and more readable to use variadic templates. [Here's an example of `is_finished_all`](https://godbolt.org/g/kARVRF).

Reply:
It'll still need tuples to be able to call op (within apply) with all the different pointers. You can't get around pointers and static sequences for this as far as I know. If I want to call a function from within apply_op, it'll need to be able to accept tuples.

Reply:
Hmm I'm not sure if I understand the constraints, but I'll be happy to try to figure out a way that doesn't involve sequences if you could elaborate 😄

Reply:
You can't do ```op(*std::get<Is>(t).data()...);``` without using tuples.

Reply:
What is `t` in this case? We should just try to limit the use of tuples unless absolutely necessary, because they often cause the code to be more complicated than necessary